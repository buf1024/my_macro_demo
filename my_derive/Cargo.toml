[package]
edition = "2021"
name = "my_derive"
version = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
proc-macro = true
doctest = false

[dependencies]
proc-macro2 = "1.0.70"
quote = "1.0.33"
syn = {version = "2.0.42", features = ["fold", "full"]}

[dev-dependencies]
trybuild = "1.0.86"


[[bin]]
name = "derive_pass"
path = "./tests/ui/derive/test_pass.rs"