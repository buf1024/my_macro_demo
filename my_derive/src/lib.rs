mod derive_getters;
use derive_getters::expand_getters;

mod attr_retry;
use attr_retry::{expand_retry, Args};

mod fn_formula;
use fn_formula::{expand_formula, FormulaArgs};

use proc_macro::TokenStream;
use syn::{parse_macro_input, DeriveInput, ItemFn};

#[proc_macro_derive(Getters, attributes(getter))]
pub fn getters(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as DeriveInput);
    let token_stream = expand_getters(input);
    token_stream
        .unwrap_or_else(|e| e.into_compile_error())
        .into()
}

#[proc_macro_attribute]
pub fn retry(attr: TokenStream, input: TokenStream) -> TokenStream {
    let item_fn = parse_macro_input!(input as ItemFn);

    let args = parse_macro_input!(attr as Args);

    let token_stream = expand_retry(args.attr, item_fn);
    token_stream
        .unwrap_or_else(|e| e.into_compile_error())
        .into()
}

#[proc_macro]
pub fn formula(input: TokenStream) -> TokenStream {
    let my_formula = parse_macro_input!(input as FormulaArgs);

    let token_stream = expand_formula(my_formula);
    token_stream
        .unwrap_or_else(|e| e.into_compile_error())
        .into()
}
