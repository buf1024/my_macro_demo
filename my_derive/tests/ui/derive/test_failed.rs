#![allow(dead_code)]

use my_derive::Getters;

#[derive(Getters)]
enum MyEnum {
    #[getter(vis=pub(crate))]
    #[getter(name = "get_fuck_data")]
    Data,
}

fn main() {}
