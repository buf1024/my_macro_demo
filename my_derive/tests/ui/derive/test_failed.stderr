error: proc-macro derive panicked
 --> tests/ui/derive/test_failed.rs:5:10
  |
5 | #[derive(Getters)]
  |          ^^^^^^^
  |
  = help: message: only allow struct filed
